src/main/java/com/google/devtools/build/lib/bazel/bzlmod/IndexRegistry.java

```diff
@@ -92,14 +106,45 @@ private String constructUrl(String base, String... segments) {
   }
 
   /** Grabs a file from the given URL. Returns {@link Optional#empty} if the file doesn't exist. */
-  private Optional<byte[]> grabFile(String url, ExtendedEventHandler eventHandler)
+  private Optional<byte[]> grabFile(
+      String url, ExtendedEventHandler eventHandler, boolean useChecksum)
+      throws IOException, InterruptedException {
+    var maybeContent = doGrabFile(url, eventHandler, useChecksum);
+    if (knownFileHashesMode == KnownFileHashesMode.USE_AND_UPDATE && useChecksum) {
+      eventHandler.post(RegistryFileDownloadEvent.create(url, maybeContent));
+    }
+    return maybeContent;
+  }
+
+  private Optional<byte[]> doGrabFile(
+      String url, ExtendedEventHandler eventHandler, boolean useChecksum)
       throws IOException, InterruptedException {
+    Optional<Checksum> checksum;
+    if (knownFileHashesMode != KnownFileHashesMode.IGNORE && useChecksum) {
+      Optional<Checksum> knownChecksum = knownFileHashes.get(url);
+      if (knownChecksum == null) {
+        // This is a new file, download without providing a checksum.
+        checksum = Optional.empty();
+      } else if (knownChecksum.isEmpty()) {
+        // The file is known to not exist, so don't attempt to download it.
```

Comment:
I can't remember whether we discussed this in the doc -- 404s can legally become present at any given time, technically. Can we handle them like we handle yanked versions, perhaps? (also, is the special handling of yanked versions in this PR? I'm not seeing it)

Reply:
We discussed this in the doc and came to the conclusion that it's necessary for reproducibility.

I think that the cleanest solution would be a new lockfile mode `update_remote` (name TBD) that attempts refetches of module files that previously 404ed and also checks the yanked versions on every change to the lockfile.

`update` would keep all of this locked to ensure a fully reproducible and offline build when the module deps aren't modified: It doesn't recheck 404s and only fetches yanked versions for those modules whose `source.json` doesn't have a hash in the lock file (that is, skip the check for everything that had been part of a valid Selection result at some point and thus either wasn't yanked or was explicitly allowed at that point).